File: app/src/main/java/com/nedaluof/qurany/ui/base/BaseActivity.java
package com.nedaluof.qurany.ui.base;

import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import java.util.concurrent.atomic.AtomicLong;

/**
 * Created by nedaluof on 6/13/2020.
 */
public class BaseActivity extends AppCompatActivity {
    private static final String KEY_ACTIVITY_ID = "KEY_ACTIVITY_ID";
    private static final AtomicLong NEXT_ID = new AtomicLong(0);
   /* private static final LongSparseArray<ConfigPersistentComponent>
            sComponentsMap = new LongSparseArray<>();

    private ActivityComponent mActivityComponent;
    private long mActivityId;*/

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        //setContentView(getLayout());
        // Create the ActivityComponent and reuses cached ConfigPersistentComponent if this is
        // being called after a configuration change.
        /*mActivityId = savedInstanceState != null ?
                savedInstanceState.getLong(KEY_ACTIVITY_ID) : NEXT_ID.getAndIncrement();

        ConfigPersistentComponent configPersistentComponent = sComponentsMap.get(mActivityId, null);

        if (configPersistentComponent == null) {
            configPersistentComponent = DaggerConfigPersistentComponent.builder()
                    .applicationComponent(MagApplication.get(this).getComponent())
                    .build();
            sComponentsMap.put(mActivityId, configPersistentComponent);
        }
        mActivityComponent = configPersistentComponent.activityComponent(new ActivityModule(this));*/
    }


    //to control the state of the Activity when the Activity
    //come in onPause() state
   /* @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putLong(KEY_ACTIVITY_ID, mActivityId);
    }*/

    /*@Override
    protected void onDestroy() {
        if (!isChangingConfigurations()) {
            sComponentsMap.remove(mActivityId);
        }
        super.onDestroy();
    }

    public ActivityComponent activityComponent() {
        return mActivityComponent;
    }*/
}

File: app/src/main/java/com/nedaluof/qurany/ui/base/BaseActivity.kt
package com.nedaluof.qurany.ui.base

import android.content.Context
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.lifecycle.ViewModel
import androidx.lifecycle.lifecycleScope
import com.nedaluof.qurany.di.LocaleManagerEntryPoint
import com.nedaluof.qurany.util.LocaleManager
import dagger.hilt.android.EntryPointAccessors
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * Created by NedaluOf on 9/11/2021.
 */
abstract class BaseActivity<DB : ViewDataBinding> : AppCompatActivity() {

  lateinit var binding: DB
  abstract val layoutId: Int
  abstract val bindingVariable: Int
  abstract fun getViewModel(): ViewModel?

  @Inject
  lateinit var localeManager: LocaleManager

  override fun attachBaseContext(newBase: Context?) {
    val localeManager = EntryPointAccessors.fromApplication(
      newBase!!,
      LocaleManagerEntryPoint::class.java
    ).localeManager
    super.attachBaseContext(newBase.let { localeManager.configureAppLocale(it) })
  }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    binding = DataBindingUtil.setContentView(this, layoutId)
    with(binding) {
      setVariable(bindingVariable, getViewModel())
      lifecycleOwner = this@BaseActivity
      executePendingBindings()
    }
  }

  fun <T> StateFlow<T>.collectFlow(data: (T) -> Unit) {
    lifecycleScope.launch {
      collect { data(it) }
    }
  }
}
